src/types.ts

```diff
@@ -0,0 +1,65 @@
+export interface Spec<T> {
+  /**
+   * An Array that lists the admissable parsed values for the env var.
+   */
+  choices?: ReadonlyArray<T>
+  /**
+   * A fallback value, which will be used if the env var wasn't specified. Providing a default effectively makes the env var optional.
+   */
+  default?: T
+  /**
+   * A fallback value to use only when NODE_ENV is not 'production'.
+   * This is handy for env vars that are required for production environments, but optional for development and testing.
+   */
+  devDefault?: T
+  /**
+   * A string that describes the env var.
+   */
+  desc?: string
+  /**
+   * An example value for the env var.
+   */
+  example?: string
+  /**
+   * A url that leads to more detailed documentation about the env var.
+   */
+  docs?: string
+}
+
+export interface ValidatorSpec<T> extends Spec<T> {
+  type: string
+  _parse: (input: string) => T
+}
+
+export interface CleanEnv {
+  /** true if NODE_ENV === 'development' */
+  readonly isDevelopment: boolean
+  readonly isDev: boolean
+
+  /** true if NODE_ENV === 'test' */
+  readonly isTest: boolean
+
+  /** true if NODE_ENV === 'production' */
+  readonly isProduction: boolean
+  readonly isProd: boolean
+}
+
+interface ReporterOptions {
+  errors: { [key: string]: Error }
```

Comment:
Because TypeScript is a bit stupid with hashmaps

```suggestion
  errors: { [key: string]: Error | undefined }
```
or,
```suggestion
  errors: Record<string, Error | undefined>
```

Reply:
I don't understand why `undefined` is necessary here– what problem does it solve? If a key is in that object, there should be a corresponding error

Reply:
Because this is wrong
<img width="598" alt="Screenshot_2020-10-09_at_16_37_56" src="https://user-images.githubusercontent.com/459267/95596686-00ad0d00-0a4e-11eb-8a2f-6e997f22f0d4.png">


Try here: https://www.typescriptlang.org/play?#code/MYewdgzgLgBCBGArAjALhgbwNoGsCmAnutAE4CWYA5gLrFTlUC+MAvJgFCPvsIoB0UPNBgB6ETAA8AWikwAhhAgBXALZCYg4WQjyYpCpW6hIsXgCZ02fET30DtWw0oAfGErAATPADMKeD8xsGJzc5gJCsGKSMvKKquqasNq6+lQwAPogJOlunj5+HtzsUQDKcmqxcEhm7MbCvADM6ABKeKAkHhKplAA0jgbO7l6+YP4AfKwcXDzV4cJR0rIKymo6iTDJcv1pmdm5wwVFUQBCAPIAcgCqJejnZMB4JDAA7iRyAA7vj+xQBF8wABF7lAJAAVCZsVrtTrdPqgwZ5EbjIzgepIAAs6CBwBB3QhmGmvHRc0i4kWsRWCQiGx0W26GSyOSG+VGhSAA


I like the `Dict<string>`-pattern.

Reply:
I tried going one further and adding a type parameter to ReporterOptions, so the keys from the specs should be inferred, and `undefined` should no longer be possible. Let me know if you see any issues with it